### PR TITLE
Line hinges

### PR DESCRIPTION
**Pull request by katulino**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Unit Tests
- [ ] Attached examples
- [ ] Own example attached to Pull Request

**Test Configuration**:
* RFEM / RSTAB / RSECTION version:
# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: includes/TypesForLines/LineMeshRefinement.js [hunk 1]
```diff
@@ -1,8 +1,163 @@
...
+* @param	{Number}			no						Index of line mesh refinement, can be undefined
+* @param	{Number}			target_length			Length of FEA element side on the line, can be undefined
+* @param	{Number}			layers					Number of layers to be affected by this refinement, can be undefined
+* @param	{Number or Array}	lines					Lines assigned to this refinement, can be undefined
+* @param	{String}			comment					Comment, can be undefined
+* @param	{Object}			params					line mesh refinement parameters, can be undefined
+* @return	{Object}			Created line mesh refinement HLF
+*/
 function LineMeshRefinement(no,
+							target_length,
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:59]: remove this argument and assing default value in function body - 0.1m

File: includes/TypesForLines/LineMeshRefinement.js [hunk 2]
```diff
@@ -1,8 +1,163 @@
...
+							layers,
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:60]: remove this argument and assing default value in function body - 2

File: includes/TypesForLines/LineMeshRefinement.js [hunk 3]
```diff
@@ -1,8 +1,163 @@
...
+							lines,
                             comment,
                             params)
 {
-    var line_mesh_refinement = engine.create_line_mesh_refinement(no);
-    set_comment_and_parameters(line_mesh_refinement, comment, params);
-    return line_mesh_refinement;
+    this.settings = createLineMeshRefinement(no, target_length, layers, lines, comment, params);
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:65]: add condition if(arguments.lenght !==0){} and put code inside this block. User could have option to create just LineMeshRefinement object and then use one function from the three main types.

File: includes/TypesForLines/LineMeshRefinement.js [hunk 4]
```diff
@@ -1,8 +1,163 @@
...
+* @param	{Number}			layers					Number of layers to be affected by this refinement, can be undefined
+* @param	{Number or Array}	lines					Lines assigned to this refinement, can be undefined
+* @return	{Object}			Created line mesh refinement
+*/
+LineMeshRefinement.prototype.Lenght = function(target_length, layers, lines) {
+	
+	this.settings.type = line_mesh_refinements.TYPE_LENGTH;
+	if (target_length != undefined) {
+		this.settings.target_length = target_length;
+	}
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:83]: else - use default value

File: includes/TypesForLines/LineMeshRefinement.js [hunk 5]
```diff
@@ -1,8 +1,163 @@
...
+
+	if (layers != undefined) {
+		this.settings.number_of_layers = layers;
+	}
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:87]: else - use default value

File: includes/TypesForLines/LineMeshRefinement.js [hunk 6]
```diff
@@ -1,8 +1,163 @@
...
+
+	if (lines != undefined) {
+		this.settings.lines = lines;
+	}
+
+	return this.settings
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:93]: Add missing semicolon

File: includes/TypesForLines/LineMeshRefinement.js [hunk 7]
```diff
@@ -1,8 +1,163 @@
...
+}
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:94]: Add missing semicolon

File: includes/TypesForLines/LineMeshRefinement.js [hunk 8]
```diff
@@ -1,8 +1,163 @@
...
+
+
+/**
+* Change line mesh refinement to type gradually changed due to distance from the line
+* @param	{Number}			gradual_rows			Number gradual rows of FEA elements, can be undefined
+* @param	{Number}			layers					Number of layers to be affected by this refinement, can be undefined
+* @param	{Number or Array}	lines					Lines assigned to this refinement, can be undefined
+* @return	{Object}			Created line mesh refinement
+*/
+LineMeshRefinement.prototype.Gradual = function(gradual_rows, layers, lines) {
```
Comment (jarabroz) [inline includes/TypesForLines/LineMeshRefinement.js:129]: USe name closer to RFEM - Gradually
no, name, lines, numberOfRows, numberOfLayers

File: includes/TypesForLines/LineHinge.js [hunk 9]
```diff
@@ -1,31 +1,160 @@
...
+/**
+* Creates line hinge
+* @class
+* @constructor
+* @param	{Number}	no				Index of line hinge, can be undefined
+* @param	{String}	comment			Comment, can be undefined
+* @param	{Object}	params			line hinge parameters, can be undefined
+* @return	{Object}	Created line hinge
+*/
 function LineHinge(no,
```
Comment (jarabroz) [inline includes/TypesForLines/LineHinge.js:74]: Add argument name.
LineHinge.prototype. = function(table_id, line, surface) {
Add argument AssignTo as two dimensional array - [[table_id1, line1, surface1],[table_id2, line2, surface2]] -> use function AssignTo in the constructor. USer get object and then modify conditions

File: includes/TypesForLines/LineHinge.js [hunk 10]
```diff
@@ -1,31 +1,160 @@
...
+/**
+* Set rotation constant rx to line hinge
+* @param	{Boolean or Float}	rx				Rotation rx (true, false, number(stiffness [Nm^2]))
+* @return	{Object}	line hinge in parent
+*/
+LineHinge.prototype.Rotation = function (rx) {
+	this.lineHinge.rotational_release_phi_x = CreateHinge(rx);
+	return this.lineHinge;
+};
+
```
Comment (jarabroz) [inline includes/TypesForLines/LineHinge.js:141]: Add function Slab-wall connection

File: includes/TypesForLines/LineHinge.js [hunk 11]
```diff
@@ -1,31 +1,160 @@
...
+
```
Comment (jarabroz) [inline includes/TypesForLines/LineHinge.js:142]: Add function for nonlinearity Fix - Force/MomentDiagram

File: examples/line_hinge_test.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+			Line(undefined, [i + j, i + j + 1]);
+		}
+		Line(undefined, [i, i + j]);
+		i = i + 3;
+	}
+
+	for (var i = 1; i <= 80; ++i) {
+		Surface(undefined, [i, i + 1, i + 2, i + 3], thickness);
+		i = i + 3;
+	};
```
Comment (jarabroz) [inline examples/line_hinge_test.js:32]: unncessary semicolon

File: examples/line_hinge_test.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+	Line(undefined, "2,3");
+	Line(undefined, "6,7");
+	Line(undefined, "7,8");
+	Line(undefined, "8,9");
+	Line(undefined, "9,6");
+	Line(undefined, "4,10");
+	Line(undefined, "1,2");
+};
```
Comment (jarabroz) [inline examples/line_hinge_test.js:40]: unncessary semicolon

File: examples/line_hinge_test.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+Member(undefined, [12,3], "", { "section_start": section_IPE240 });
+Member(undefined, [4,12], "", { "section_start": section_IPE240 });
+Member(undefined, [12,16], "", { "section_start": section_IPE180 });
+//   ################# Nodal supports
+//   ##### Constructors:
+// Empty
+var LH1 = new LineHinge();
+LH1.Translation(true,true,true);
+LH1.Rotation(false);
+LH1.AssignTo(1, 1, 1)
```
Comment (jarabroz) [inline examples/line_hinge_test.js:51]: Missing semicolon - but it will be moved to LineHinge definition

File: examples/line_hinge_test.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+var material_steel = Material(undefined, "S235 | EN 1993-1-1:2005-05");
+var material_concrete = Material(undefined, "C12/15 | EN 1992-1-1:2004/A1:2014");
+
+var section_params = { "shear_stiffness_deactivated": true };
+var section_IPE240 = Section(undefined, "IPE 240", material_steel, "", section_params);
+var section_IPE180 = Section(undefined, "IPE 180", material_steel, "", section_params);
+
+if (RFEM) {
+	var thickness = Thickness(undefined, "", material_concrete, 120mm);
+};
```
Comment (jarabroz) [inline examples/line_hinge_test.js:12]: Unnecessary semicolon

File: examples/line_hinge_test.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+for (var i = 0; i < 20; ++i) {
```
Comment (jarabroz) [inline examples/line_hinge_test.js:13]: Why we need to have 20 surfaces?

File: examples/line_hinge_test.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+		}
+		Line(undefined, [i, i + j]);
+		i = i + 3;
+	}
+
+	for (var i = 1; i <= 80; ++i) {
+		Surface(undefined, [i, i + 1, i + 2, i + 3], thickness);
+		i = i + 3;
+	};
+	Line(undefined, "2,3");
```
Comment (jarabroz) [inline examples/line_hinge_test.js:33]: Why?

File: examples/line_hinge_test.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+	Line(undefined, "6,7");
+	Line(undefined, "7,8");
+	Line(undefined, "8,9");
+	Line(undefined, "9,6");
+	Line(undefined, "4,10");
+	Line(undefined, "1,2");
+};
+
+Member(undefined, [12,3], "", { "section_start": section_IPE240 });
```
Comment (jarabroz) [inline examples/line_hinge_test.js:42]: Why?

File: examples/line_hinge_test.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+Member(undefined, [4,12], "", { "section_start": section_IPE240 });
+Member(undefined, [12,16], "", { "section_start": section_IPE180 });
+//   ################# Nodal supports
```
Comment (jarabroz) [inline examples/line_hinge_test.js:45]: There are not any nodal supports

File: examples/line_hinge_test.js [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+var LH5 = new LineHinge();
+LH5.TranslationX(450);
+LH5.TranslationY(false);
+LH5.TranslationZ(320);
+
+var params = {
+	user_defined_name_enabled : true,
+	name : " user name "
+}
```
Comment (jarabroz) [inline examples/line_hinge_test.js:76]: Missing semicolon

File: examples/line_hinge_test.js [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+new LineHinge(undefined, "comment", params)
```
Comment (jarabroz) [inline examples/line_hinge_test.js:78]: Missing semicolon

File: examples/line_hinge_test.js [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+NR.Elements();
+NR.settings.elements_finite_elements = 15;
+NR.settings.lines = [1,2,3,6];
+
+var NR1 = new LineMeshRefinement();
+NR1.Gradual();
+NR1.SetLines([10,13,17]);
+
+var NR2 = new LineMeshRefinement()
```
Comment (jarabroz) [inline examples/line_hinge_test.js:92]: Missing semicolon

File: examples/line_hinge_test.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+NR2.Gradual(7,4);
+
+var NR3 = new LineMeshRefinement()
```
Comment (jarabroz) [inline examples/line_hinge_test.js:95]: Missing semicolon

**Review by jarabroz — CHANGES_REQUESTED** [review]

File: includes/TypesForLines/LineMeshRefinement.js [hunk 24]
```diff
@@ -1,8 +1,163 @@
...
+
+
+/**
+* Change line mesh refinement to type based on number of elements on line
+* @param	{Number}			elements			Number of FEA elements on the line, can be undefined
+* @param	{Number}			layers				Number of layers to be affected by this refinement, can be undefined
+* @param	{Number or Array}	lines				Lines assigned to this refinement, can be undefined
+* @return	{Object}			Created line mesh refinement
+*/
+LineMeshRefinement.prototype.Elements = function(elements, layers, lines) {
```
[the remark this answers is not in the document]
Comment (katulino) [reply includes/TypesForLines/LineMeshRefinement.js:104]: LineMeshRefinement.prototype.NumberFiniteElements = function(numberOfFiniteElements, numberOfLayers, lines) {

File: includes/TypesForLines/LineMeshRefinement.js (hunk 3 above)
Comment (katulino) [reply includes/TypesForLines/LineMeshRefinement.js]: it is already possible

**Review by jarabroz — APPROVED** [review]

**Review by jarabroz — APPROVED** [review]

**Pull request merged**